src/modules/mc_pos_control/mc_pos_control_main.cpp

```diff
@@ -770,2524 +369,510 @@ MulticopterPositionControl::task_main_trampoline(int argc, char *argv[])
+		// timed out - periodic check for _task_should_exit
+		if (pret == 0) {
+			// Go through the loop anyway to copy manual input at 50 Hz.
+		}
 
-		/* reset slew rate */
-		_vel_sp_prev(2) = _vel(2);
-		_user_intention_z = brake;
-	}
+		// this is undesirable but not much we can do
+		if (pret < 0) {
+			warn("poll error %d, %d", pret, errno);
+			continue;
+		}
 
-	_user_intention_z = intention;
+		poll_subscriptions();
 
-	/*
-	 * apply acceleration depending on state
-	 */
-	if (_user_intention_z == brake) {
+		parameters_update(false);
+
+		hrt_abstime t = hrt_absolute_time();
+		const float dt = t_prev != 0 ? (t - t_prev) / 1e6f : 0.004f;
+		t_prev = t;
 
-		/* limit jerk when braking to zero */
-		float jerk = (_acceleration_z_max_up.get() - _acceleration_state_dependent_z) / _dt;
+		// set dt for control blocks
+		setDt(dt);
 
-		if (jerk > _manual_jerk_limit_z) {
-			_acceleration_state_dependent_z = _manual_jerk_limit_z * _dt + _acceleration_state_dependent_z;
+		if (_control_mode.flag_armed) {
+			// as soon vehicle is armed check for flighttask
+			start_flight_task();
+			// arm hysteresis prevents vehicle to takeoff
+			// before propeller reached idle speed.
+			_arm_hysteresis.set_state_and_update(true);
 
 		} else {
-			_acceleration_state_dependent_z = _acceleration_z_max_up.get();
+			// disable flighttask
+			_flight_tasks.switchTask(FlightTaskIndex::None);
+			// reset arm hysteresis
+			_arm_hysteresis.set_state_and_update(false);
 		}
-	}
-
-	if (_user_intention_z == acceleration) {
-		_acceleration_state_dependent_z = (max_acceleration - _acceleration_z_max_down.get()) * fabsf(
-				stick_z) + _acceleration_z_max_down.get();
-	}
-}
 
-void
-MulticopterPositionControl::set_manual_acceleration_xy(matrix::Vector2f &stick_xy)
-{
+		// check if any task is active
+		if (_flight_tasks.isAnyTaskActive()) {
 
-	/*
-	 * In manual mode we consider four states with different acceleration handling:
-	 * 1. user wants to stop
-	 * 2. user wants to quickly change direction
-	 * 3. user wants to accelerate
-	 * 4. user wants to decelerate
-	 */
+			// setpoints from flighttask
+			vehicle_local_position_setpoint_s setpoint;
 
-	/* get normalized stick input vector */
-	matrix::Vector2f stick_xy_norm = (stick_xy.length() > 0.0f) ? stick_xy.normalized() : stick_xy;
-	matrix::Vector2f stick_xy_prev_norm = (_stick_input_xy_prev.length() > 0.0f) ? _stick_input_xy_prev.normalized() :
-					      _stick_input_xy_prev;
+			// update task
+			if (!_flight_tasks.update()) {
+				// FAILSAFE
+				// Task was not able to update correctly. Do Failsafe.
+				setpoint.x = setpoint.y = setpoint.z = NAN;
+				setpoint.vx = setpoint.vy = setpoint.vz = NAN;
+				setpoint.thrust[0] = setpoint.thrust[1] = setpoint.thrust[2] = NAN;
 
-	/* check if stick direction and current velocity are within 60angle */
-	const bool is_aligned = (stick_xy_norm * stick_xy_prev_norm) > 0.5f;
+				if (PX4_ISFINITE(_states.velocity(2))) {
```

Comment:
move this stuff outside. this requires to check for landdetector and  if in air, then slowly descend, if landed, then just go idle